types/node/globals.d.ts

```diff
@@ -238,111 +238,39 @@ declare var exports: any;
 
 // Buffer class
 type BufferEncoding = "ascii" | "utf8" | "utf-8" | "utf16le" | "ucs2" | "ucs-2" | "base64" | "latin1" | "binary" | "hex";
-interface Buffer extends Uint8Array {
+
+interface Buffer {
     constructor: typeof Buffer;
```

Comment:
Do we care about `Buffer` being subclassed? Specifying `constructor` prevents `Buffer` subclasses from having differing construct signatures. I'm aware this is an existing declaration, but it seems like it could be problematic.

Reply:
See previous discussion, I'm personally against defining it as it is a VERY niche use case, but it would be a breaking change.

Reply:
Until all classes T have `.constructor` as a type T, this is important for me to be able to correctly type `qs` (something i'm gradually exploring). It'd be appreciated if it wasn't removed.

Reply:
> Until all classes T have .constructor as a type T, this is important for me to be able to correctly type qs (something i'm gradually exploring). It'd be appreciated if it wasn't removed.

We will probably never do this as it would break subclassing where the subclass has a differing construct signature:
```ts
class SuperClass {
  a: Foo;
  constructor(a: Foo) {
    this.a = a;
  }
}

class SubClass extends SuperClass {
  b: Bar;
  constructor(opts: { a: Foo, b: Bar }) {
    super(opts.a);
    this.b = opts.b;
  }
}

interface SuperClass {
    constructor: typeof SuperClass;
}

interface SubClass {
    //    ^^^^^^^^
    // error: class 'SubClass' incorrectly extends base class 'SuperClass'.
    //  Types of property 'constructor' are incompatible.
    //    Type 'typeof SubClass' is not assignable to type 'typeof SuperClass'.
    //      Types of parameters 'opts' and 'a' are incompatible.
    //        Type 'string' is not assignable to type '{ a: string; b: string; }'.
    constructor: typeof SubClass;
}
```

Playground Link: https://www.typescriptlang.org/play/index.html#code/MYGwhgzhAEDKCuAHApgJwMLitA3gKGmjAC5oIAXVASwDsBzAbgOmAHsaLV5hzXUAKEmUq06ASlzNC5ABZUIAOjDQAvESaEAvnm15QkGAgBGmA9GQAPcshoATQ0jSns+QkdKdRGlu07deAqyI5BCkOEQeIvQANNDuwtT00JoSroRkjoHBimBi3tJyikaq0EEhCkbe2rq01qgAZmDAyHCZzjBphGwclP58pOQAniis9a0oGFgQTDU0dY3NrSZTkuk+PVw8-dBDI2PG7TNAA

Reply:
I guess i don’t understand why, because the constructor property has no relationship to super, and I’d expect with the interface approach to have to subtract out the constructor, or union it.

Reply:
If you type `.constructor` as `typeof SuperClass`, you might expect to be able to do this:
```ts
class SuperClass {
  a: string;
  constructor(a: string) {
    this.a = a;
  }
  clone() {
    return new this.constructor(this.a);
  }
}
```
However, this isn't sound as any subclass could have an incompatible construct signature.

Reply:
Then it should be the union of all constructor types, such that if they’re incompatible you need an explicit cast to call it.

Reply:
You can't have a "union of all constructor types" when typing `constructor` of `SuperClass`, because you can't know all of the possible subclasses at that time.

For your case, the more sound approach would be to do something like:
```js
const { Buffer } = require("buffer");
const { Buffer: SaferBuffer } = require("safer-buffer");
var isBuffer = function isBuffer(obj) {
  return Buffer.isBuffer(obj) || SaferBuffer.isBuffer(obj);
}
```

I could see a *possible* future mechanism for `this.constructor` that returned the static side of the containing class *without* call or construct signatures (but retaining the apparent type of `Function`). Then in `qs` you would do `(obj: (Buffer | Value) & { constructor?: { isBuffer?(obj: any): boolean; } }) => boolean;`

Reply:
In qs, we can’t add those deps since it would increase bundle size for everyone whether they’re using Buffer or not.

Reply:
(though I suppose `SaferBuffer.isBuffer` is just `Buffer.isBuffer`)

Reply:
Is it strictly necessary to enforce `Buffer`'s constructor is `Buffer` then? From a type-perspective the assumption is that if you are using the `Buffer` type from `@types/node` then you should be a `Buffer` (hand-wavy type-casts like `x as unknown as Buffer` aside). If you can't import or reference `Buffer` at runtime, your only type safety would be a duck-typed best-fit scenario:
```js
var isBuffer = function (obj) {
  ...
  if (obj.constructor && typeof obj.constructor.isBuffer === "function" && obj.constructor.isBuffer.length >= 1) {
    try {
      const result = obj.constructor.isBuffer(obj);
      return typeof result === "boolean" && result; // pretty sure this could be a buffer...
    }
    catch {
      return false; // definitely not a buffer
    }
  }
  return false; // no way to know  
}
```

Reply:
@ljharb: I filed https://github.com/microsoft/TypeScript/issues/32452 as a suggestion for improving the type of `constructor`. This includes a possible workaround for the subclassing issue that could be employed today: 

```ts
type StaticMembers<TClass extends Function> = Pick<TClass, keyof TClass> & Function;
interface Buffer {
  constructor: StaticMembers<typeof Buffer>;
}
```